File: qqbot-persistence/src/main/java/com/mieai/qqbot/persistence/audit/AuditLog.java
package com.mieai.qqbot.persistence.audit;

import java.time.Instant;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

/** Sanitized administrator mutation record; request bodies are never stored. */
public record AuditLog(
        UUID id,
        Optional<String> actorUsername,
        String action,
        String resourcePath,
        int outcomeStatus,
        Optional<String> remoteAddress,
        Optional<String> traceId,
        Instant createdAt) {

    public AuditLog {
        Objects.requireNonNull(id, "id must not be null");
        actorUsername = requireOptional(actorUsername, "actorUsername", 64);
        action = requireText(action, "action", 16);
        resourcePath = requireText(resourcePath, "resourcePath", 512);
        if (outcomeStatus < 100 || outcomeStatus > 599) {
            throw new IllegalArgumentException("outcomeStatus must be a valid HTTP status");
        }
        remoteAddress = requireOptional(remoteAddress, "remoteAddress", 128);
        traceId = requireOptional(traceId, "traceId", 128);
        Objects.requireNonNull(createdAt, "createdAt must not be null");
    }

    private static String requireText(String value, String name, int maximumLength) {
        Objects.requireNonNull(value, name + " must not be null");
        if (value.isBlank() || value.length() > maximumLength
                || value.codePoints().anyMatch(Character::isISOControl)) {
            throw new IllegalArgumentException(name + " is invalid");
        }
        return value;
    }

    private static Optional<String> requireOptional(
            Optional<String> value, String name, int maximumLength) {
        Objects.requireNonNull(value, name + " must not be null");
        return value.map(candidate -> requireText(candidate, name, maximumLength));
    }
}

File: qqbot-persistence/src/main/kotlin/com/mieai/qqbot/persistence/audit/AuditLog.kt
package com.mieai.qqbot.persistence.audit

import java.time.Instant
import java.util.UUID

/** Sanitized administrator mutation record; request bodies are never stored. */
data class AuditLog(
    val id: UUID,
    val actorUsername: String?,
    val action: String,
    val resourcePath: String,
    val outcomeStatus: Int,
    val remoteAddress: String?,
    val traceId: String?,
    val createdAt: Instant,
) {
    init {
        validateNullable(actorUsername, "actorUsername", 64)
        requireText(action, "action", 16)
        requireText(resourcePath, "resourcePath", 512)
        require(outcomeStatus in 100..599) { "outcomeStatus must be a valid HTTP status" }
        validateNullable(remoteAddress, "remoteAddress", 128)
        validateNullable(traceId, "traceId", 128)
    }

    companion object {
        private fun requireText(value: String, name: String, maximumLength: Int) {
            require(value.isNotBlank() && value.length <= maximumLength &&
                value.codePoints().noneMatch { Character.isISOControl(it) }) {
                "$name is invalid"
            }
        }

        private fun validateNullable(value: String?, name: String, maximumLength: Int) {
            value?.let { requireText(it, name, maximumLength) }
        }
    }
}
